src/packagedcode/go_sum.py

```diff
@@ -0,0 +1,133 @@
+#  OR CONDITIONS OF ANY KIND, either express or implied. No content created from
+#  ScanCode should be considered or used as legal advice. Consult an Attorney
+#  for any legal advice.
+#  ScanCode is a free software code scanning tool from nexB Inc. and others.
+#  Visit https://github.com/nexB/scancode-toolkit/ for support and download.
+
+from __future__ import absolute_import
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import io
+import itertools
+import logging
+import re
+
+import attr
+
+
+"""
+https://blog.golang.org/using-go-modules
+
+For example:
+
+golang.org/x/text v0.0.0-20170915032832-14c0d48ead0c h1:qgOY6WgZO...
+golang.org/x/text v0.0.0-20170915032832-14c0d48ead0c/go.mod h1:Nq...
+rsc.io/quote v1.5.2 h1:w5fcysjrx7yqtD/aO+QwRjYZOKnaM9Uh2b40tElTs3...
+rsc.io/quote v1.5.2/go.mod h1:LzX7hefJvL54yjefDEDHNONDjII0t9xZLPX...
+rsc.io/sampler v1.3.0 h1:7uVkIFmeBqHfdjD+gZwtXXI+RODJ2Wc4O7MPEh/Q...
+rsc.io/sampler v1.3.0/go.mod h1:T1hPZKmBbMNahiBKFy5HrXp6adAjACjK9...
+
+"""
+
+"""
+go.sum file contains certain things in 2 formats.
+
+For example:
+github.com/BurntSushi/toml v0.3.1 h1:WXkYYl6Yr3qBf1K79EBnL4mak0OimBfB0XUf9Vl28OQ=
+github.com/BurntSushi/toml v0.3.1/go.mod h1:xHWCNGjB5oqiDr8zfno3MHue2Ht5sIBksp03qcyfWMU=
+
+
+>>> ob = GoSum()
+>>> p = ob.parse_dep_type2('github.com/BurntSushi/toml v0.3.1 h1:WXkYYl6Yr3qBf1K79EBnL4mak0OimBfB0XUf9Vl28OQ=')
+>>> assert p.group('namespace') == ('github.com/BurntSushi')
+>>> assert p.group('name') == ('toml')
+>>> assert p.group('version') == ('v0.3.1')
+>>> assert p.group('checksum') == ('WXkYYl6Yr3qBf1K79EBnL4mak0OimBfB0XUf9Vl28OQ=')
+
+>>> p = ob.parse_dep_type1('github.com/BurntSushi/toml v0.3.1/go.mod h1:xHWCNGjB5oqiDr8zfno3MHue2Ht5sIBksp03qcyfWMU=')
+>>> assert p.group('namespace') == ('github.com/BurntSushi')
+>>> assert p.group('name') == ('toml')
+>>> assert p.group('version') == ('v0.3.1')
+>>> assert p.group('checksum') == ('xHWCNGjB5oqiDr8zfno3MHue2Ht5sIBksp03qcyfWMU=')
+"""
+
+
+TRACE = False
+
+logger = logging.getLogger(__name__)
+
+if TRACE:
+    import sys
+    logging.basicConfig(stream=sys.stdout)
+    logger.setLevel(logging.DEBUG)
+
+
+@attr.s()
+class GoSum(object):
+    # Regex expressions to parse different types of dependency
+    # dep_type1 example: github.com/BurntSushi/toml v0.3.1 h1:WXkYY....
+    parse_dep_type1 = re.compile(
+        r'(?P<namespace>(.*))'
+        r'\/'
+        r'(?P<name>[^\s]*)'
+        r'(\s)*'
+        r'(?P<version>(.*))'
+        r'/go.mod(\s)*h1:'
+        r'(?P<checksum>(.*))'
+    ).match
+
+    # dep_type2 example: github.com/BurntSushi/toml v0.3.1/go.mod h1:xHWCN....
+    parse_dep_type2 = re.compile(
+        r'(?P<namespace>(.*))'
+        r'/'
+        r'(?P<name>[^\s]*)'
+        r'(\s)*'
+        r'(?P<version>(.*))'
+        r'(\s)*h1:'
+        r'(?P<checksum>(.*))'
+    ).match
+
+    @classmethod
+    def parse_gosum(cls, location):
```

Comment:
What I meant earlier with a class method is for it to be a factory to return GoSum objects that would have actual real fields such as:
```
@attr.s()
class GoSum(object):
    namespace = attr.ib(default=None)
    name = attr.ib(default=None)
    version = attr.ib(default=None)
```
and not return a list of tuples

Reply:
@pombredanne currently my go.sum parser return list of list which contain namespace, name, version in order like:
if this is the go.sum file:
```
github.com/BurntSushi/toml v0.3.1 h1:WXkYYl6Yr3qBf1K79EBnL4mak0OimBfB0XUf9Vl28OQ=
github.com/BurntSushi/toml v0.3.1/go.mod h1:xHWCNGjB5oqiDr8zfno3MHue2Ht5sIBksp03qcyfWMU=
github.com/cznic/golex v0.0.0-20181122101858-9c343928389c h1:G8zTsaqyVfIHpgMFcGgdbhHSFhlNc77rAKkhVbQ9kQg=
github.com/cznic/golex v0.0.0-20181122101858-9c343928389c/go.mod h1:+bmmJDNmKlhWNG+gwWCkaBoTy39Fs+bzRxVBzoTQbIc=
```
then the return value will be `[['github.com/BurntSushi', 'toml', 'v0.3.1'], ['github.com/cznic', 'golex', 'v0.0.0-20181122101858-9c343928389c']]`

You told me to use classmethod factory, so what output format you want from classmethod factory?? plz give a example

Reply:
I was suggesting to return a list of Gosum objects... I find it always better and easier to read and understand when the code uses objects with named attributes rather than to use nested lists.

NB: I pushed a small doc header to your branch

Reply:
@pombredanne I have updated my PR. I think I have done changes according to you but I am not sure fully. Please review the PR again :)